File: src/server/cache-version.ts
import { getDb } from "@/db/client";
import { widgetCache } from "@/db/schema";
import { getPref, setPref } from "./config-repo";

/**
 * Bump whenever any widget's Data payload shape changes. The cache is
 * disposable by design (everything is re-fetchable), so a mismatch wipes it —
 * no per-widget payload migrations.
 */
export const CACHE_VERSION = 4;

export async function ensureCacheVersion(): Promise<void> {
  const stored = await getPref("cacheVersion", "");
  if (stored === String(CACHE_VERSION)) return;
  await getDb().delete(widgetCache);
  await setPref("cacheVersion", String(CACHE_VERSION));
}

File: internal/dashboard/cache_version.go
package dashboard

import (
	"context"
	"fmt"
)

// CacheVersion identifies the shape of every widget's cached Data payload.
// Bump it whenever any widget's Data payload shape changes; a mismatch
// against the stored pref wipes the disposable widget cache on boot.
const CacheVersion = 1

const cacheVersionPref = "cacheVersion"

// EnsureCacheVersion wipes the disposable widget cache when the payload shape
// version changed. Wipe-then-stamp: a crash between the two just re-wipes next boot.
//wails:ignore
func (s *Service) EnsureCacheVersion() error {
	ctx := context.Background()
	stored, err := s.store.Pref(ctx, cacheVersionPref, "")
	if err != nil {
		return err
	}
	if stored == fmt.Sprint(CacheVersion) {
		return nil
	}
	if err := s.store.CacheWipe(ctx); err != nil {
		return err
	}
	return s.store.SetPref(ctx, cacheVersionPref, fmt.Sprint(CacheVersion))
}
